### ISSUE TITLE
Feature Request:  Import units and settings from previous version

### ISSUE THREAD

**Issue by BLR-IIC** [MegaMek/megamek#4849]
This came up on the BattleTech forums.  Instead of manually copying and pasting data files and settings, I think it would be great if there was a tool built into MegaMek that could import these files semi-automatically from a previous version of the software.  I'd like to see MekHQ do this too with campaigns.

**Comment by Sleet01** [comment]
I've actually taken a look at this.  Copying specific customized files over from one installed dir to another shouldn't be too difficult.  The major problem point seems to be that specific fields in the config and unit files (and also MM save games!) routinely change _names_, not that the data is necessarily incompatible.  We may fail to read some subset of the fields, or completely fail to read a file, and then the outcomes are not well defined.  As one example, I corrected a spelling error between two 0.49.X versions, which rendered the save games incompatible between those two versions because the mis-spelled name had been used explicitly in the save game files!

The simplest approach would be to create translation layer definitions for each `release` -> `release'` transition, and attempt to apply every existing layer in order against the file that the user wants to load.  Once a file loads cleanly, great: just keep applying translation layers to each new converted file until you get to the current version.  If the file never loads, just tell the user it's incompatible.

That said, having worked on version compatibility stuff for other software, it would take a _lot_ of work to set up the initial translation system and layer definitions.  I've managed to do very limited save game loading from 0.48.0 or 0.49.X into later versions of 0.49.Y, and it took a lot of trial and error.  Before we even begin we need to:

A) create a tool that can programmatically spit out all the loadable fields whose names _or_ types have changed between two releases, and all the affected file names within the release;
B) create standardized type conversions (e.g. `str` -> `float`, or in the case or 0.49.14's munition types, `'long` -> `Enum`) that the conversion layer will be able to call based on the defined input and output;
C) create (or source) a number of files from which to create unit tests to validate the functioning of the translation code.

And then _finally_ the work of creating the translation layer definitions (e.g. something like
```
0.49.13 -> 0.49.14: 
    *.mtf:
        TransportBayBA:
            name: TransportBayBAIS,
            type: str:str
```
etc., etc. could begin.

In this model, there's a specific file for every release and its subsequent release, up to the latest.  Then we have to create a new translation definition for every file and field that has changed between releases... for every release, forever.  This is likely a small but not insignificant additional overhead we'd have to commit to for... a long time.

tl;dr: making a tool to just copy a subset of files over from old install to new install is easy; making it useful and error-resistant is pretty hard, and not to be undertaken lightly.

**Comment by HammerGS** [comment]
Which is the very reason the user data stuff has had limited development.

**Comment by SuperStucco** [comment]
Part of the translation, the XML files, should be doable via XSL which can handle everything from changing tags to re-formatting data.  I've routinely used this to translate between XML and other XML, HTML, and CSV output.  That was in C# though, which may have more extensive built in support for applying XSL templates.
